Feature/schema mutation

janardhan1993 opened the pull request:
We can set schema now through gql

mutation {
schema {
	name: string @index
    initial_release_date: date @index
    director.film:uid @reverse
}
}

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/710)
<!-- Reviewable:end -->

manishrjain commented:
Reviewed 25 of 25 files at r1.
Review status: all files reviewed at latest revision, 28 unresolved discussions, some commit checks failed.

---

*[client/client.go, line 143 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe0JyZ4ra_w_2dHZcN:-Kfe0Jy_yTpaznnB5bHH:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L143)):*
> ```Go
> 	if typ == types.UidID && schema.Index {
> 		// index on uid type
> 		return x.Errorf("Index not allowed on predicate of type uid on predicate %s", schema.Predicate)
> ```

100 chars.

---

*[client/client.go, line 182 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe1D8qxJevPfVle1Sb:-Kfe0wdOKn5dD9xr4DEh:b-rcomzr) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L182)):*
> ```Go
> 	pending int
> 
> 	nquads chan nquadOp
> ```

nquadOp to have a wrapper struct; which can take both the nquad and the schema update.

---

*[client/client.go, line 183 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe0lByApGu0vkVkG9w:-Kfe0lByApGu0vkVkG9x:b1948ma) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L183)):*
> ```Go
> 
> 	nquads chan nquadOp
> 	schema chan []*graphp.SchemaUpdate
> ```

schema chan graphp.SchemaUpdate

---

*[client/client.go, line 185 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe33GTXKOXxXof9vH_:-Kfe33GTXKOXxXof9vHa:b-nvlnh3) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L185)):*
> ```Go
> 	schema chan []*graphp.SchemaUpdate
> 	dc     graphp.DgraphClient
> 	wg     sync.WaitGroup
> ```

wg.Add(..)

---

*[client/client.go, line 186 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe34kv3p9asJJTOpOF:-Kfe34kv3p9asJJTOpOG:bfjoq95) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L186)):*
> ```Go
> 	dc     graphp.DgraphClient
> 	wg     sync.WaitGroup
> 	swg    sync.WaitGroup
> ```

No need for this.

---

*[client/client.go, line 233 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe2zLug3wvgZox87tl:-Kfe2zLug3wvgZox87tm:b28d1rx) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L233)):*
> ```Go
> 	req := new(Req)
> 
> 	for s := range batch.schema {
> ```

Have a separate goroutine with a for loop; and inside select over the schema channel and default. In schema, push to array. In default, check the size, acquire locks, and send to Dgraph. Otherwise, sleep for 10ms.

---

*[client/client.go, line 243 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe36WVyPI6nV2ucQLn:-Kfe36WVyPI6nV2ucQLo:byifezs) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L243)):*
> ```Go
> 		batch.request(req)
> 	}
> 	batch.swg.Done()
> ```

batch.wg.Done()

---

*[client/client.go, line 260 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe396wdvC2j12XNrIM:-Kfe396wdvC2j12XNrIN:bmrkyvz) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L260)):*
> ```Go
> 		go bm.makeRequests()
> 	}
> 	bm.swg.Add(1)
> ```

bm.wg.Add(1)

---

*[client/client.go, line 274 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe3CkWAJT7atamvQlo:-Kfe3CkWAJT7atamvQlp:bpdicpr) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L274)):*
> ```Go
> }
> 
> func (batch *BatchMutation) AddSchema(updates []*graphp.SchemaUpdate) error {
> ```

AddSchema(update graphp.SchemaUpdate)

---

*[client/client.go, line 303 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe3H1bWYC7AU1DiHJf:-Kfe3H1bWYC7AU1DiHJg:bqb7ydy) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L303)):*
> ```Go
> }
> 
> func (batch *BatchMutation) FlushSchema() {
> ```

Merge with Flush().

---

*[client/client_test.go, line 133 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe1foGSqLjoiP0enJA:-Kfe1foGSqLjoiP0enJB:b-6wa7fl) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client_test.go#L133)):*
> ```Go
> 		for i := 1; i <= 3; i++ {
> 			go func() {
> 				select {
> ```

for loop around select.

---

*[cmd/dgraph/main.go, line 268 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe45XlXqAQKcetnDph:-Kfe45XlXqAQKcetnDpi:b-oik1yv) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L268)):*
> ```Go
> 			schema.Tokenizer = []string{tok.Default(typ).Name()}
> 		}
> 		if !schema.Index {
> ```

if len(schema.Tokenizer) == 0 .. else if schema.Index { ... }

---

*[cmd/dgraph/main.go, line 272 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe4OLZD6L9BtXvyqe8:-Kfe4OLZD6L9BtXvyqe9:bj5k36d) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L272)):*
> ```Go
> 		}
> 		// check for valid tokeniser types and duplicates
> 		var seen = make(map[string]bool)
> ```

Don't remove dups, error out. Let the user deal with it.

---

*[cmd/dgraph/main.go, line 463 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe4e0X705ozmuxZvNl:-Kfe4e0X705ozmuxZvNm:bgog33v) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L463)):*
> ```Go
> 	var allocIdsStr map[string]string
> 	// If we have mutations, run them first.
> 	if res.Mutation != nil && (len(res.Mutation.Set) > 0 ||
> ```

if res.Mutation != nil &&
  (len(Set) > 0 || len(Del) > 0 || len(schema) > 0) { .. }

---

*[cmd/dgraph/main.go, line 628 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe53xNq3A78gY8T3eb:-Kfe53xNq3A78gY8T3ec:bz63b2w) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L628)):*
> ```Go
> 	// Mutations are sent as part of the mutation object
> 	if req.Mutation != nil && (len(req.Mutation.Set) > 0 ||
> 		len(req.Mutation.Del) > 0 || len(res.Mutation.Schema) > 0) {
> ```

Actually, move it to a small function which can check this for you.

---

*[cmd/dgraph/main_test.go, line 294 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe5e8yXe1HCOXGUVgR:-Kfe5e8yXe1HCOXGUVgS:b3wppu9) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main_test.go#L294)):*
> ```Go
> 
> 	// add index to name
> 	err = runMutation(s)
> ```

On query receive and parsed; if mutation has schema update; set context to background. Otherwise, keep it to the deadline.

---

*[cmd/dgraphloader/main.go, line 97 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe6hivaL60m1F-uIUG:-Kfe6hiwMOE06WL2Bz0-:b-9dttjb) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraphloader/main.go#L97)):*
> ```Go
> 
> 	}
> 	batch.FlushSchema()
> ```

remove this.

---

*[cmd/dgraphloader/main.go, line 183 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe78QB6NOE1x7Jvqv2:-Kfe78QB6NOE1x7Jvqv3:buos9nu) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraphloader/main.go#L183)):*
> ```Go
> 	x.AssertTrue(len(filesList) > 0)
> 	if len(*schemaFile) > 0 {
> 		processSchemaFile(*schemaFile, batch)
> ```

time.Sleep(time.Second) // 1 second.

---

*[posting/index.go, line 263 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe84-S8y3HJH4fpJ1T:-Kfe84-S8y3HJH4fpJ1U:b-51dnck) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/posting/index.go#L263)):*
> ```Go
> 	}
> 
> 	for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
> ```

So, you'll be evicting all the Pls from memory, before rebuilding the index; or before deleting the index.

Also, switch lhmap so it's one lhmap per group.

---

*[protos/graphp/graphresponse.proto, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe9t9g43YG6Lhc-qV1:-Kfe9t9g43YG6Lhc-qV2:b-y6c2x1) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/protos/graphp/graphresponse.proto#L53)):*
> ```protobuf
>     string predicate = 1;
>     uint32 value_type = 2;
>     bool index = 3;
> ```

use enum for index and reverse.

---

*[schema/parse.go, line 94 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeA1htBuiQ7E7W3g-y:-KfeA1htBuiQ7E7W3g-z:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L94)):*
> ```Go
> 					return nil, x.Errorf("Cannot reverse for non-UID type")
> 				}
> 				return &graphp.SchemaUpdate{Predicate: predicate, ValueType: uint32(t), Reverse: true}, nil
> ```

100 chars.

---

*[schema/parse.go, line 102 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeA2_-L-kTVgpN09pO:-KfeA2_-L-kTVgpN09pP:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L102)):*
> ```Go
> 					return nil, err
> 				} else {
> 					return &graphp.SchemaUpdate{Predicate: predicate, ValueType: uint32(t), Index: true, Tokenizer: tokenizer}, nil
> ```

100 chars.

---

*[schema/parse.go, line 157 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeACnlsjDaFSY2xvLL:-KfeACnlsjDaFSY2xvLM:b-ri51xt) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L157)):*
> ```Go
> 
> // Parse parses a schema string and returns the schema representation for it.
> func Parse(s string) ([]*graphp.SchemaUpdate, error) {
> ```

[]graphp.SchemaUpdate

---

*[schema/parse_test.go, line 121 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeAZdJ6RnhFmVFBdD6:-KfeAZdJ6RnhFmVFBdD7:b-7h2on) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse_test.go#L121)):*
> ```Go
> func TestParse(t *testing.T) {
> 	reset()
> 	schemas, err := Parse("age:int @index name:string")
> ```

Have a delim. A "dot + newline". It's better to be restrictive to avoid confusion later.

---

*[schema/schema.go, line 135 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeC4JnHoQ8TsoBoqNL:-KfeC4JnHoQ8TsoBoqNM:bb23he3) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/schema.go#L135)):*
> ```Go
> 	}
> 	// create a copy
> 	sc := *schema
> ```

Don't create a copy.

---

*[worker/backup.go, line 193 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeDlRPaZUPzr3HtJkP:-KfeDlRPaZUPzr3HtJkQ:b-rvisay) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/backup.go#L193)):*
> ```Go
> 
> 	// Use a goroutine to convert typesp.Schema to string
> 	chs := make(chan *skv, 1000)
> ```

Add a TODO to ensure that data is flushed to disk before iterating for backup. And then, we only look at the disk; without worrying about what's in memory.

This should be done via the scheduler.

---

*[worker/draft.go, line 490 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeEKQYn2GpKomBuu6E:-KfeEKQYn2GpKomBuu6F:b-s2k371) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/draft.go#L490)):*
> ```Go
> 				// Wait for applied watermark to reach till previous index
> 				// All mutations before this should use old schema and after this
> 				// should use new schema
> ```

Add a TODO to do this in scheduler.

---

*[worker/index.go, line 101 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeGgFCT-0X7ZID_Z9V:-KfeGgFDI_bFciLXQnXN:b-x1nhfp) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/index.go#L101)):*
> ```Go
> }
> 
> func (n *node) applyAllMarks(ctx context.Context, lastIndex uint64) error {
> ```

waitForAppliedMark

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/710)*
<!-- Sent from Reviewable.io -->

janardhan1993 commented:
Review status: 9 of 29 files reviewed at latest revision, 29 unresolved discussions, some commit checks failed.

---

*[client/client.go, line 143 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe0JyZ4ra_w_2dHZcN:-KfjMFNHqlglku0JO_Ed:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L143)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

---

*[client/client.go, line 182 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe1D8qxJevPfVle1Sb:-KfjMJjnJXVhAG2gJYVz:b-do9drd) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L182)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

nquadOp to have a wrapper struct; which can take both the nquad and the schema update.
</blockquote></details>

We get schema in a separate channel so that schema could be applied periodically without waiting for batch size.

---

*[client/client.go, line 183 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe0lByApGu0vkVkG9w:-KfjMYS26fIm5xU4Juxr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L183)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

schema chan graphp.SchemaUpdate
</blockquote></details>

Done.

---

*[client/client.go, line 185 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe33GTXKOXxXof9vH_:-KfjM_KUlACB1nc_1YMf:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L185)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

wg.Add(..)
</blockquote></details>

Done.

---

*[client/client.go, line 186 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe34kv3p9asJJTOpOF:-KfjMa1MPPghQID3zq1n:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L186)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

No need for this.
</blockquote></details>

Done.

---

*[client/client.go, line 233 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe2zLug3wvgZox87tl:-KfjMfugMDbNLwky46TD:b7akqrx) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L233)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Have a separate goroutine with a for loop; and inside select over the schema channel and default. In schema, push to array. In default, check the size, acquire locks, and send to Dgraph. Otherwise, sleep for 10ms.
</blockquote></details>

We sleep in dgraphloader after adding schema mutations, so we can ignore the lock for now

---

*[client/client.go, line 243 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe36WVyPI6nV2ucQLn:-KfjMuoVtMWVOlzhrN0T:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L243)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

batch.wg.Done()
</blockquote></details>

Done.

---

*[client/client.go, line 260 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe396wdvC2j12XNrIM:-KfjMw0OHsSsAhlVF5rg:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L260)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

bm.wg.Add(1)
</blockquote></details>

Done.

---

*[client/client.go, line 274 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe3CkWAJT7atamvQlo:-KfjMxJpDrMJ6b8Q9Y3M:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L274)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

AddSchema(update graphp.SchemaUpdate)
</blockquote></details>

Done.

---

*[client/client.go, line 303 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe3H1bWYC7AU1DiHJf:-KfjMyPX17SpxPpY-oKo:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client.go#L303)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Merge with Flush().
</blockquote></details>

Done.

---

*[client/client_test.go, line 133 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe1foGSqLjoiP0enJA:-KfjN-KnBQNsPrqGnKrN:bsd4ssr) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/client/client_test.go#L133)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

for loop around select.
</blockquote></details>

There wasn't any significant difference

---

*[cmd/dgraph/main.go, line 268 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe45XlXqAQKcetnDph:-KfjN4A4NKmWwxIuXQPD:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L268)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

if len(schema.Tokenizer) == 0 .. else if schema.Index { ... }
</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 272 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe4OLZD6L9BtXvyqe8:-KfjN5Pcvt4nIHULcS7L:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L272)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't remove dups, error out. Let the user deal with it.
</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 463 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe4e0X705ozmuxZvNl:-KfjN7Bh5Z2eFwVqkPWj:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L463)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

if res.Mutation != nil &&
  (len(Set) > 0 || len(Del) > 0 || len(schema) > 0) { .. }

</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 628 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe53xNq3A78gY8T3eb:-KfjN8P-OtruYDtjOlG-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main.go#L628)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Actually, move it to a small function which can check this for you.
</blockquote></details>

Done.

---

*[cmd/dgraph/main_test.go, line 294 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe5e8yXe1HCOXGUVgR:-KfjNDA3AYADTvSLtMdc:b2vk50p) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraph/main_test.go#L294)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

On query receive and parsed; if mutation has schema update; set context to background. Otherwise, keep it to the deadline.
</blockquote></details>

Done in main.go, have added a timeout of 100 hours to simulate context.background instead of creating new context(we have values stored in context already)

---

*[cmd/dgraphloader/main.go, line 97 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe6hivaL60m1F-uIUG:-KfjORaetWr4eR_C3y9H:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraphloader/main.go#L97)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

remove this.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 183 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe78QB6NOE1x7Jvqv2:-KfjOSSyMqT8FN5yHYSO:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/cmd/dgraphloader/main.go#L183)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

time.Sleep(time.Second) // 1 second.
</blockquote></details>

Done.

---

*[posting/index.go, line 263 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe84-S8y3HJH4fpJ1T:-KfjNUfWzzVjueIwSezX:b-iqi4w2) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/posting/index.go#L263)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

So, you'll be evicting all the Pls from memory, before rebuilding the index; or before deleting the index.

Also, switch lhmap so it's one lhmap per group.
</blockquote></details>

Done. One lhmap per group , i will raise a separate PR for that

---

*[protos/graphp/graphresponse.proto, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-Kfe9t9g43YG6Lhc-qV1:-KfjN_dE_sSTwfGhOliU:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/protos/graphp/graphresponse.proto#L53)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

use enum for index and reverse.
</blockquote></details>

Done.

---

*[schema/parse.go, line 94 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeA1htBuiQ7E7W3g-y:-KfjNbk61b-jPFJpIXeA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L94)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

---

*[schema/parse.go, line 102 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeA2_-L-kTVgpN09pO:-KfjNcjhHt_nOkWD-y4_:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L102)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

---

*[schema/parse.go, line 157 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeACnlsjDaFSY2xvLL:-KfjNfCv0DNiQuQOU3Ip:bxbjz3o) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse.go#L157)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

[]graphp.SchemaUpdate
</blockquote></details>

Mutation proto buff needs []*graphp.SchemaUpdate

---

*[schema/parse_test.go, line 121 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeAZdJ6RnhFmVFBdD6:-KfjNnwz2wkdW4Z91nmZ:brcxg2i) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/parse_test.go#L121)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Have a delim. A "dot + newline". It's better to be restrictive to avoid confusion later.
</blockquote></details>

Will raise a separate PR. Should it be a dot + newline ? rdf has dot as delimiter

---

*[schema/schema.go, line 135 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeC4JnHoQ8TsoBoqNL:-KfjO-dQ2vodv-ULWlYP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/schema/schema.go#L135)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't create a copy.
</blockquote></details>

Done.

---

*[worker/backup.go, line 193 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeDlRPaZUPzr3HtJkP:-KfjO6BX3ZgWFx54VyjI:b-wjp92k) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/backup.go#L193)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a TODO to ensure that data is flushed to disk before iterating for backup. And then, we only look at the disk; without worrying about what's in memory.

This should be done via the scheduler.
</blockquote></details>

Done.Added a TODO in scheduler

---

*[worker/draft.go, line 490 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeEKQYn2GpKomBuu6E:-KfjODCNeYCPqO7pD0qZ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/draft.go#L490)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a TODO to do this in scheduler.
</blockquote></details>

Done.

---

*[worker/index.go, line 32 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfjOGvceqKk6MtM7m8L:-KfjOGvceqKk6MtM7m8M:bnli7ot) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/index.go#L32)):*
> ```Go
> 		// wait for synced watermarks if we delete through mutations, but
> 		// it would use by lhmap
> 		go posting.DeleteIndex(ctx, attr)
> ```

Just to avoid issues, making it blocking call until we have scheduler

---

*[worker/index.go, line 101 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KfeGgFCT-0X7ZID_Z9V:-KfjOFh24WQ0YpfScVsC:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/338af9ec02b62c46b70585d3bacc6cd21347cee3/worker/index.go#L101)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

waitForAppliedMark
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/710)*
<!-- Sent from Reviewable.io -->

F21 commented:
If we update the schema using this feature, will it automatically rebuild indexes or will we have to call `admin/index?attr=yourpredicate`?

janardhan1993 commented:
@F21  It would automatically reindex. In face we will remove the reindex endpoint.

Motakjuq commented:
@janardhan1993 do I need to define all the schema every time I perform a mutation? Does it exist a way to get the current schema?

janardhan1993 commented:
@Motakjuq  No you needn't define schema every time. Once you change it it's persisted. 

you could send the following to query endpoint to get the current schema 

schema {
			index
			reverse
			type
			tokenizer
		}

or the following to get schema for pred1, pred2 and pred3 edge

schema(pred: [pred1, pred2, pred3]) {
			index
			reverse
			type
			tokenizer
		}

Motakjuq commented:
Thx, I get it

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 19 of 19 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[client/client.go, line 245 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftNw3jlteA-qGyXVLM:-KftNw3jlteA-qGyXVLN:b-ba7pov) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/client/client.go#L245)):*
> ```Go
> 				batch.request(req)
> 			}
> 			time.Sleep(10 * time.Millisecond)
> ```

So, get start time, and deduct that duration against batch.request duration; so we're always waiting only 10ms, not more.

---

*[cmd/dgraph/main.go, line 465 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftOjmrTr2mfkbB_3is:-KftOjmrTr2mfkbB_3it:bxlxo33) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L465)):*
> ```Go
> 	var allocIdsStr map[string]string
> 	// If we have mutations, run them first.
> 	if res.Mutation != nil && isNonEmptyGQLMutation(res.Mutation) {
> ```

hasGQLOps(res.Mutation)

---

*[cmd/dgraph/main.go, line 468 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftPWzUYrqmO2-oM5X4:-KftPWzVDTrA8ntxsJED:b-ni9e0) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L468)):*
> ```Go
> 		if len(res.Mutation.Schema) > 0 {
> 			// setting a higher value to ensure schema mutations are completed
> 			ctx, _ = context.WithTimeout(ctx, 100*time.Hour)
> ```

So, use context.Background upfront. And for non-schema ops, change that to WithTimeout. Make sure that you defer cancel.

---

*[cmd/dgraph/main.go, line 593 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftP0QR5EAakQXJeaQ5:-KftP0QSkv51LPfY-8H2:btih2jd) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L593)):*
> ```Go
> }
> 
> func isNonEmptyGraphMutation(mu *graphp.Mutation) bool {
> ```

hasGraphOps

---

*[cmd/dgraph/main_test.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftPkXIeG95Mg_YS38U:-KftPkXIeG95Mg_YS38V:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main_test.go#L173)):*
> ```Go
> 	schema.ParseBytes([]byte(""), 1)
> 	expected := map[string]*typesp.Schema{
> 		"name": {Tokenizer: []string{"term", "exact"}, ValueType: uint32(types.StringID), Directive: typesp.Schema_INDEX},
> ```

100 chars.

---

*[cmd/dgraphloader/main.go, line 183 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQEn_hD9fVYMcplgn:-KftQEn_hD9fVYMcplgo:bto18f5) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraphloader/main.go#L183)):*
> ```Go
> 	}
> 	// wait for schema changes to be done before starting mutations
> 	time.Sleep(1)
> ```

time.Sleep(time.Second)

---

*[schema/parse.go, line 41 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQOhMv7TO9QZSxDdN:-KftQOhMv7TO9QZSxDdO:b-pgmhqu) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L41)):*
> ```Go
> func From(s *graphp.SchemaUpdate) typesp.Schema {
> 	if s.Directive == graphp.SchemaUpdate_REVERSE {
> 		return typesp.Schema{ValueType: s.ValueType, Directive: typesp.Schema_REVERSE}
> ```

Split into multiple lines:

typesp.Schema{
  ValueType: ..,
  Directive: ...
}

---

*[schema/parse.go, line 43 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQcdwS8BWnBwkogxJ:-KftQcdwS8BWnBwkogxK:b-61j0xh) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L43)):*
> ```Go
> 		return typesp.Schema{ValueType: s.ValueType, Directive: typesp.Schema_REVERSE}
> 	} else if s.Directive == graphp.SchemaUpdate_INDEX {
> 		return typesp.Schema{ValueType: s.ValueType, Directive: typesp.Schema_INDEX,
> ```

same here. split.

---

*[schema/parse.go, line 100 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQgZL05oAURnn53kE:-KftQgZL05oAURnn53kF:bb9opfc) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L100)):*
> ```Go
> 					return nil, x.Errorf("Cannot reverse for non-UID type")
> 				}
> 				return &graphp.SchemaUpdate{Predicate: predicate, ValueType: uint32(t),
> ```

split them into multiple lines.

---

*[schema/parse.go, line 109 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQiWf6RvDsiE2Aizt:-KftQiWf6RvDsiE2Aizu:b9w1yic) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L109)):*
> ```Go
> 					return nil, err
> 				} else {
> 					return &graphp.SchemaUpdate{Predicate: predicate, ValueType: uint32(t),
> ```

and here as well.

---

*[schema/parse.go, line 156 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQmcaeJ0YxsbFh9xg:-KftQmcaeJ0YxsbFh9xh:b-77q3gh) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L156)):*
> ```Go
> 			return tokenizers, x.Errorf("Invalid tokenizer %s", next.Val)
> 		}
> 		if _, ok := seen[tokenizer.Name()]; !ok {
> ```

if _, found := ..; found { return error }
tokenizers = append(..)

---

*[schema/parse_test.go, line 108 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftR34IPoSPtpLGV1bI:-KftR34IPoSPtpLGV1bJ:byn8w8i) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse_test.go#L108)):*
> ```Go
> 	require.NoError(t, ReloadData("testfiles/test_schema_index4", 1))
> 	checkSchema(t, State().get(1).predicate, []nameType{
> 		{"name", &typesp.Schema{ValueType: uint32(types.StringID), Tokenizer: []string{"exact"}, Directive: typesp.Schema_INDEX}},
> ```

1 kv pair per line.

---

*[worker/mutation.go, line 117 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftRhtN2a7nEEzvQWvB:-KftRhtN2a7nEEzvQWvC:bd44cbl) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/worker/mutation.go#L117)):*
> ```Go
> 				return err
> 			}
> 		} else if (current.Directive == typesp.Schema_REVERSE) != (old.Directive == typesp.Schema_REVERSE) {
> ```

line length.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-:-KftRozV0m8t2CEGOEfL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

janardhan1993 commented:
Review status: 21 of 28 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[client/client.go, line 245 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftNw3jlteA-qGyXVLM:-KftnH8eGSc8axsk6-ZA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/client/client.go#L245)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

So, get start time, and deduct that duration against batch.request duration; so we're always waiting only 10ms, not more.
</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 465 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftOjmrTr2mfkbB_3is:-KftnMrZp2HWbZGq-Svr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L465)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

hasGQLOps(res.Mutation)
</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 468 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftPWzUYrqmO2-oM5X4:-KftnNenA8EnIrjtGlVW:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L468)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

So, use context.Background upfront. And for non-schema ops, change that to WithTimeout. Make sure that you defer cancel.
</blockquote></details>

Done.

---

*[cmd/dgraph/main.go, line 593 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftP0QR5EAakQXJeaQ5:-KftnOWoYI3NpBIGJN26:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main.go#L593)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

hasGraphOps
</blockquote></details>

Done.

---

*[cmd/dgraph/main_test.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftPkXIeG95Mg_YS38U:-KftnPp8m-KFeDfDNYbD:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraph/main_test.go#L173)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

---

*[cmd/dgraphloader/main.go, line 183 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQEn_hD9fVYMcplgn:-KftnR2OEHF9Yj-FPNte:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/cmd/dgraphloader/main.go#L183)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

time.Sleep(time.Second)
</blockquote></details>

Done.

---

*[schema/parse.go, line 41 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQOhMv7TO9QZSxDdN:-KftnSABmnFIF15dYIkX:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L41)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Split into multiple lines:

typesp.Schema{
  ValueType: ..,
  Directive: ...
}
</blockquote></details>

Done.

---

*[schema/parse.go, line 43 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQcdwS8BWnBwkogxJ:-KftnSqeVIP9uPtVQfh_:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L43)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

same here. split.
</blockquote></details>

Done.

---

*[schema/parse.go, line 100 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQgZL05oAURnn53kE:-KftnTgHp0krmMoBOMqY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L100)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

split them into multiple lines.
</blockquote></details>

Done.

---

*[schema/parse.go, line 109 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQiWf6RvDsiE2Aizt:-KftnUXUBqjS0TFKIzWn:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L109)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

and here as well.
</blockquote></details>

Done.

---

*[schema/parse.go, line 156 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftQmcaeJ0YxsbFh9xg:-KftnVPmkxwyfvr8zvRH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse.go#L156)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

if _, found := ..; found { return error }
tokenizers = append(..)
</blockquote></details>

Done.

---

*[schema/parse_test.go, line 108 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftR34IPoSPtpLGV1bI:-KftnWJlNYA2lHxlbzki:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/schema/parse_test.go#L108)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

1 kv pair per line.
</blockquote></details>

Done.

---

*[worker/mutation.go, line 117 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/710#-KftRhtN2a7nEEzvQWvB:-Kftn_M1STrlU2-b6hm2:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/8766782703e8f5ee73c461469869b8d43f555d60/worker/mutation.go#L117)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

line length.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/710)*
<!-- Sent from Reviewable.io -->
